lib/Fhp/Action/SendSEPATransfer.php

```diff
@@ -35,21 +47,52 @@ class SendSEPATransfer extends BaseAction
      */
     public static function create(SEPAAccount $account, string $painMessage): SendSEPATransfer
     {
-        if (preg_match('/xmlns="(.*?)"/', $painMessage, $match) === false) {
+        if (preg_match('/xmlns="(.*?)"/', $painMessage, $xmlns) === false) {
             throw new \InvalidArgumentException('xmlns not found in the PAIN message');
         }
+
+        // Check whether the PAIN message contains the correct message root
+        if (strstr($painMessage, '<CstmrCdtTrfInitn>') === false) {
+            throw new \InvalidArgumentException('Pain message contains a wrong message root');
+        }
+
+        // Check whether the PAIN message contains multiple or only one transfer, should match <NbOfTxs>xx</NbOfTxs> in the XML
+        $nbOfTxs = substr_count($painMessage, '<CdtTrfTxInf>');
+        if ($nbOfTxs === 0) {
+            throw new \InvalidArgumentException('PAIN message contains no credit transfer transactions');
+        }
+
+        $ctrlSum = null;
+        if (preg_match('@<GrpHdr>.*<CtrlSum>(?<ctrlsum>[.0-9]+)</CtrlSum>.*</GrpHdr>@s', $painMessage, $matches) === 1) {
+            $ctrlSum = $matches['ctrlsum'];
```

Comment:
floatval()

Reply:
This is copied from SendSEPADirectDebit.php and should probably be updated there, too?

Reply:
Yes I'd say so. Otherwise the property type of the class member is wrong and the conversion to float would happen only due to the argument type declaration on `Btg::create()`, which is quite unexpected behavior. I can see no reason why casting to a float here already would be wrong.